src/magic_folder/magicpath.py

```diff
@@ -3,11 +3,40 @@
 
 from pyutil.assertutil import precondition, _assert
 
-def path2magic(path):
-    return re.sub(u'[/@]',  lambda m: {u'/': u'@_', u'@': u'@@'}[m.group(0)], path)
 
-def magic2path(path):
-    return re.sub(u'@[_@]', lambda m: {u'@_': u'/', u'@@': u'@'}[m.group(0)], path)
+def mangle_path_segments(segments):
```

Comment:
I see why this and `unmangle_relative_path` are not mirrors (based on how they are used), but I wonder if that will be confusing down the line. Additionally, I wonder if the fact that this combines segments will be surprising.

Reply:
One way to address this would be to have a `relative_to` function, that does `segmentsFrom` and then joints them with `/`.

Reply:
Reading through some of the tests and things, it seems there are enough places that start with a relative path, that I think pushing the joining code to the caller that does `.segementsFrom` probably does make sense.

Reply:
should i just put the original `magicpath` code back, then?

Reply:
At least part of my motivation for changing it is that we should only ever be "mangling" relative paths, and the only FilePath API for that returns list-of-strings...maybe an even-different API would make it harder to screw up?

Something like `mangle_relative_to(base, path)` where base, path are both FilePaths and `unmangle_relative_to(base, mangled)` returns a FilePath (an absolute one, but starting at "base")

Reply:
> should i just put the original magicpath code back, then?

If we decide to go with that API, it probably makes sense, though the formatting of `unmangle_relative_path` is definitely more readable than the one-line version.

> At least part of my motivation for changing it is that we should only ever be "mangling" relative paths, and the only FilePath API for that returns list-of-strings...maybe an even-different API would make it harder to screw up?

I definitely understand the desire for an interface that is hard to get wrong. I'm not sure how to balance that desire against having an unsymmetrical API and all the places where we call the API with an already relative path where we are now calling `.split`. 

> Something like mangle_relative_to(base, path) where base, path are both FilePaths and unmangle_relative_to(base, mangled) returns a FilePath (an absolute one, but starting at "base")

There are a number of places where we interact with mangled names where we don't have or want and absolute unmangled path, so I don't think these make sense.

Reply:
I only see "split" calls in tests .. so a different way to mitigate that would be to change how the `strategies` produce relative paths.

Reply:
After writing https://github.com/LeastAuthority/magic-folder/pull/356#discussion_r638951794, one thought that occurred to me is to have:
```python
def mangle_relative_to(path: FilePath, base: FilePath) -> unicode:
    ...
def mangle_relative_path(path: FilePath) -> unicode:
    ...
def unmangle_relative_path(mangled: unicode) -> FilePath:
    ...
```
and both the single argument versions could check that `.isabs` is `False` before proceeding. (Note that we'd want the first two implemented via the same internal function, or one in terms of the other)

Reply:
> I only see "split" calls in tests .. so a different way to mitigate that would be to change how the `strategies` produce relative paths.

I can imagine wanting to add the ability to access files in snapshots directly, which would also have the same kind of issue.

Reply:
> There are a number of places where we interact with mangled names where we don't have or want and absolute unmangled path, so I don't think these make sense.

I guess there's kind of two uses here: in one sense the mangled-relative-name is just the "name" or "label" of a Snapshot; in another sense it is turned into a local filesystem path. _Conceptually_ we probably shouldn't ever care that the "name" happens to be the (mangled) path relative to the magic-folder *except* when we're about to open a file (for reading or writing).

So, where are we using an unmangled relative path? Perhaps those places should just not depend on that knowledge at all and simply use it like a label? (I think we do need to receive unmangled relative paths in APIs .. e.g. the "add a snapshot" API)